File: python/staircase.py
#!/usr/bin/env python

import time
import sys

def staircase(message, delay):
    staircase = []

    for letter in message:
        staircase.append(letter)
        print(''.join(staircase))
        sys.stdout.flush() # Just gonna put it here because time.sleep() doesn't work for git bash

        time.sleep(delay)

    while len(staircase) != 0:
        del staircase[-1]
        print(''.join(staircase))
        sys.stdout.flush() # Just gonna put it here because time.sleep() doesn't work for git bash

        time.sleep(delay)

if __name__ == '__main__':
    message = input('Message: ')
    delay = 0

    try:
        delay = float(input('Delay: '))
    except ValueError:
        delay = 0

    staircase(message, delay)

File: cpp/staircase.cpp
#include <iostream>
#include<windows.h>

using namespace std;

void staircase(string message, int delay) {
    string staircase = "";

    for(int i = 0; i < message.length(); i++) {
        staircase += message[i];
        cout << staircase << endl;
        Sleep(delay);
    }

    while(staircase.length() != 0) {
        staircase.erase(staircase.length() - 1);
        cout << staircase << endl;
        Sleep(delay);
    }
}

int main() {
    string message = "";
    cout << "Message: ";
    getline(cin, message);

    string delayString = "";
    double delay = 0;
    cout << "Delay: ";
    getline(cin, delayString);

    if(!delayString.empty())
        delay = stod(delayString);
    else
        delay = 0;

    staircase(message, delay * 1000);
}
